packages/jest-parallel/src/index.js

```diff
@@ -0,0 +1,182 @@
+/**
+ * Copyright (c) 2017-present, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ */
+
+'use strict';
+
+const mergeStream = require('merge-stream');
+const path = require('path');
+
+const JestWorker = require('./worker');
+
+import type {JestApi, JestFarmOptions} from './types';
+import type {Readable} from 'stream';
+
+/**
+ * The Jest farm is a class that allows you to queue methods across multiple
+ * child processes, in order to parallelize work. This is done by providing an
+ * absolute path to a module that will be loaded on each of the child
+ * processes, and bridged to the main process.
+ *
+ * Bridged methods are specified by using the "exposedMethods" property of the
+ * options "object". This is an array of strings, where each of them corresponds
+ * to the exported name in the loaded module.
+ *
+ * You can also control the amount of workers by using the "workers" property of
+ * the "options" object, and the settings passed to fork the process through
+ * the "forkOptions" property.
+ *
+ * Queueing calls can be done in two ways:
+ *   - Standard method: calls will be redirected to the first available worker,
+ *     so they will get executed as soon as they can.
+ *
+ *   - Sticky method: if a "getWorkerKey" method is provided within the config,
+ *     the resulting string of this method will be used as a key. Everytime this
+ *     key is returned, it is guaranteed that your job will be processed by the
+ *     same worker. This is specially useful if your workers are caching
+ *     results.
+ */
+class JestFarm {
+  _aggregagedStdout: Readable;
+  _aggregagedStderr: Readable;
+  _api: JestApi;
+  _ending: boolean;
+  _hashes: {[string]: JestWorker};
+  _options: JestFarmOptions;
+  _workers: Array<JestWorker>;
+
+  constructor(options: JestFarmOptions) {
+    const api = this._makeCall.bind(this, '');
+    const aggregatedStdout = mergeStream();
+    const aggregatedStderr = mergeStream();
+    const workers = [];
+
+    if (!path.isAbsolute(options.workerPath)) {
```

Comment:
could we instead just run "require.resolve" here ourselves? It will throw if the module isn't found, and it allows people to pass `my-worker-package` which may be a form of a "plugin API" for workers some day.

Reply:
We can't. `require.resolve` is relative to the file calling it. If we call it ourselves, all paths will always be relative to `jest-parallel` itself, which is undesirable.

Reply:
Let's call require.resolve anyway as it'll work with packages (like 'my-worker-package') and throw otherwise?

Reply:
Would https://github.com/sindresorhus/resolve-from help?

Reply:
Nah, I think that's going too far. It should either be a pre-resolved module or a node_module.

Reply:
`resolve-from` would work, but you need to go one level up the stack to check who called into `jest-parallel`. I really dislike that approach.